File: src/utils/embedding.ts
/**
 * Embedding Client (v3 — Multi-provider)
 *
 * Supports multiple embedding providers for flexible deployment:
 *   - openai: OpenAI API (text-embedding-3-small/large)
 *   - ollama: Self-hosted Ollama (nomic-embed-text, mxbai-embed-large)
 *   - vllm: Self-hosted vLLM with embedding models
 *   - tei: HuggingFace Text Embeddings Inference (recommended for on-prem)
 *   - local: Deterministic pseudo-embeddings (dev/test only)
 *
 * Provider is selected via EMBEDDING_PROVIDER env var.
 * Self-hosted providers connect to EMBEDDING_URL.
 */

import { getConfig } from '../config/index.js';
import { createChildLogger } from './logger.js';

const logger = createChildLogger({ module: 'embedding' });

export class EmbeddingClient {
  private model: string;
  private dimensions: number;
  private provider: string;
  private selfHostedUrl: string;

  constructor() {
    const config = getConfig();
    this.model = config.EMBEDDING_MODEL;
    this.dimensions = config.EMBEDDING_DIMENSIONS;
    this.provider = config.EMBEDDING_PROVIDER;
    this.selfHostedUrl = (config as any).EMBEDDING_URL ?? '';
  }

  async embed(text: string): Promise<number[]> {
    const results = await this.embedBatch([text]);
    return results[0];
  }

  async embedBatch(texts: string[]): Promise<number[][]> {
    if (texts.length === 0) return [];

    logger.debug({ count: texts.length, provider: this.provider }, 'Generating embeddings');

    const truncated = texts.map(t => t.substring(0, 8000));
    const batchSize = this.provider === 'openai' ? 100 : 32;
    const allEmbeddings: number[][] = [];

    for (let i = 0; i < truncated.length; i += batchSize) {
      const batch = truncated.slice(i, i + batchSize);
      const embeddings = await this.callProvider(batch);
      allEmbeddings.push(...embeddings);
    }

    return allEmbeddings;
  }

  private async callProvider(texts: string[]): Promise<number[][]> {
    const maxRetries = 3;
    let lastError: Error | null = null;

    for (let attempt = 0; attempt < maxRetries; attempt++) {
      try {
        switch (this.provider) {
          case 'openai':
            return await this.callOpenAI(texts);
          case 'ollama':
            return await this.callOllama(texts);
          case 'vllm':
            return await this.callVLLM(texts);
          case 'tei':
            return await this.callTEI(texts);
          case 'local':
            return this.generateLocalEmbeddings(texts);
          default:
            throw new Error(`Unknown embedding provider: ${this.provider}`);
        }
      } catch (error) {
        lastError = error as Error;
        logger.warn({ attempt, err: error, provider: this.provider }, 'Embedding call failed, retrying');
        await new Promise(resolve => setTimeout(resolve, Math.pow(2, attempt) * 1000));
      }
    }

    throw new Error(`Embedding failed after ${maxRetries} attempts: ${lastError?.message}`);
  }

  // ─── OpenAI ──────────────────────────────────────────────────────────────

  private async callOpenAI(texts: string[]): Promise<number[][]> {
    const config = getConfig();
    const response = await fetch('https://api.openai.com/v1/embeddings', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${config.OPENAI_API_KEY}`,
      },
      body: JSON.stringify({
        model: this.model,
        input: texts,
        dimensions: this.dimensions,
      }),
    });

    if (!response.ok) {
      const err = await response.text();
      throw new Error(`OpenAI embeddings failed: ${response.status} ${err}`);
    }

    const data = await response.json() as any;
    return data.data.map((d: any) => d.embedding);
  }

  // ─── Ollama ──────────────────────────────────────────────────────────────

  private async callOllama(texts: string[]): Promise<number[][]> {
    // Ollama processes one text at a time via /api/embeddings
    const results: number[][] = [];

    for (const text of texts) {
      const response = await fetch(`${this.selfHostedUrl}/api/embeddings`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ model: this.model, prompt: text }),
      });

      if (!response.ok) {
        throw new Error(`Ollama embedding failed: ${response.status}`);
      }

      const data = await response.json() as any;
      results.push(data.embedding);
    }

    return results;
  }

  // ─── vLLM ────────────────────────────────────────────────────────────────

  private async callVLLM(texts: string[]): Promise<number[][]> {
    // vLLM uses OpenAI-compatible API at /v1/embeddings
    const response = await fetch(`${this.selfHostedUrl}/v1/embeddings`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ model: this.model, input: texts }),
    });

    if (!response.ok) {
      throw new Error(`vLLM embedding failed: ${response.status}`);
    }

    const data = await response.json() as any;
    return data.data.map((d: any) => d.embedding);
  }

  // ─── HuggingFace Text Embeddings Inference (TEI) ─────────────────────────

  private async callTEI(texts: string[]): Promise<number[][]> {
    // TEI uses POST /embed with { inputs: [...] }
    const response = await fetch(`${this.selfHostedUrl}/embed`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ inputs: texts, truncate: true }),
    });

    if (!response.ok) {
      throw new Error(`TEI embedding failed: ${response.status}`);
    }

    const data = await response.json() as any;
    // TEI returns array of arrays directly
    return data;
  }

  // ─── Local (dev/test only) ───────────────────────────────────────────────

  private generateLocalEmbeddings(texts: string[]): number[][] {
    return texts.map(text => {
      const embedding = new Array(this.dimensions).fill(0);
      for (let i = 0; i < text.length && i < this.dimensions; i++) {
        embedding[i % this.dimensions] += text.charCodeAt(i) / 1000;
      }
      const norm = Math.sqrt(embedding.reduce((s: number, v: number) => s + v * v, 0));
      return norm > 0 ? embedding.map((v: number) => v / norm) : embedding;
    });
  }

  getModelName(): string { return this.model; }
  getDimensions(): number { return this.dimensions; }
}

File: go/internal/ingestion/embedding.go
package ingestion

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"math"
	"net/http"
	"os"
	"runtime"
	"time"
)

// EmbeddingClient handles vector embedding generation from multiple providers.
type EmbeddingClient struct {
	provider   string
	model      string
	dimensions int
	url        string
	apiKey     string
	numThread  int
	client     *http.Client
}

// NewEmbeddingClient creates an embedding client from environment config.
func NewEmbeddingClient() *EmbeddingClient {
	// runtime.NumCPU reflects the cgroup-visible CPU count, which is what local
	// inference should be sized to. EMBEDDING_NUM_THREADS overrides it.
	threads := envIntOr("EMBEDDING_NUM_THREADS", runtime.NumCPU())

	return &EmbeddingClient{
		provider:   envOr("EMBEDDING_PROVIDER", "local"),
		model:      envOr("EMBEDDING_MODEL", "nomic-embed-text"),
		dimensions: envIntOr("EMBEDDING_DIMENSIONS", 768),
		url:        envOr("EMBEDDING_URL", "http://localhost:11434"),
		apiKey:     os.Getenv("OPENAI_API_KEY"),
		numThread:  threads,
		client:     &http.Client{Timeout: 60 * time.Second},
	}
}

// Model returns the configured model name.
func (e *EmbeddingClient) Model() string { return e.model }

// Embed generates an embedding for a single text.
func (e *EmbeddingClient) Embed(ctx context.Context, text string) ([]float64, error) {
	results, err := e.EmbedBatch(ctx, []string{text})
	if err != nil {
		return nil, err
	}
	if len(results) == 0 {
		return nil, fmt.Errorf("no embedding returned")
	}
	return results[0], nil
}

// EmbedBatch generates embeddings for multiple texts.
func (e *EmbeddingClient) EmbedBatch(ctx context.Context, texts []string) ([][]float64, error) {
	if len(texts) == 0 {
		return nil, nil
	}

	// Truncate long texts
	for i := range texts {
		if len(texts[i]) > 8000 {
			texts[i] = texts[i][:8000]
		}
	}

	var (
		embeddings [][]float64
		err        error
	)
	switch e.provider {
	case "openai":
		embeddings, err = e.callOpenAI(ctx, texts)
	case "ollama":
		embeddings, err = e.callOllama(ctx, texts)
	case "tei":
		embeddings, err = e.callTEI(ctx, texts)
	case "local":
		embeddings = e.generateLocal(texts)
	default:
		return nil, fmt.Errorf("unsupported embedding provider %q", e.provider)
	}
	if err != nil {
		return nil, err
	}
	if len(embeddings) != len(texts) {
		return nil, fmt.Errorf("embedding provider returned %d vectors for %d inputs", len(embeddings), len(texts))
	}
	for i, vector := range embeddings {
		if len(vector) != e.dimensions {
			return nil, fmt.Errorf("embedding %d has %d dimensions; expected %d", i, len(vector), e.dimensions)
		}
		for _, value := range vector {
			if math.IsNaN(value) || math.IsInf(value, 0) {
				return nil, fmt.Errorf("embedding %d contains a non-finite value", i)
			}
		}
	}
	return embeddings, nil
}

func (e *EmbeddingClient) callOpenAI(ctx context.Context, texts []string) ([][]float64, error) {
	if e.apiKey == "" {
		return nil, fmt.Errorf("OPENAI_API_KEY is required for the openai embedding provider")
	}
	body := map[string]any{
		"model":      e.model,
		"input":      texts,
		"dimensions": e.dimensions,
	}
	data, err := json.Marshal(body)
	if err != nil {
		return nil, fmt.Errorf("encode openai request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://api.openai.com/v1/embeddings", bytes.NewReader(data))
	if err != nil {
		return nil, fmt.Errorf("create openai request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+e.apiKey)

	resp, err := e.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("openai embed: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(io.LimitReader(resp.Body, 64<<20))
	if err != nil {
		return nil, fmt.Errorf("openai read response: %w", err)
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("openai %d: %s", resp.StatusCode, string(respBody[:min(200, len(respBody))]))
	}

	var result struct {
		Data []struct {
			Embedding []float64 `json:"embedding"`
		} `json:"data"`
	}
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("openai decode: %w", err)
	}

	embeddings := make([][]float64, len(result.Data))
	for i, d := range result.Data {
		embeddings[i] = d.Embedding
	}
	return embeddings, nil
}

// callOllama embeds a batch via Ollama's /api/embed endpoint.
//
// Two details matter for latency. First, /api/embed accepts an array and
// returns all vectors in one round trip, unlike the older /api/embeddings which
// takes a single prompt. Second, num_thread must be set explicitly: inside a
// container llama.cpp reads the host's CPU count, so on a 4-core LXC guest it
// defaults to 16 threads and thrashes. Measured on the test box, a short embed
// went from ~8s at the default to ~57ms with num_thread matching the real core
// count.
func (e *EmbeddingClient) callOllama(ctx context.Context, texts []string) ([][]float64, error) {
	body := map[string]any{
		"model":   e.model,
		"input":   texts,
		"options": map[string]any{"num_thread": e.numThread},
	}
	data, err := json.Marshal(body)
	if err != nil {
		return nil, fmt.Errorf("encode ollama request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, e.url+"/api/embed", bytes.NewReader(data))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := e.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("ollama embed: %w", err)
	}
	defer resp.Body.Close()

	payload, err := io.ReadAll(io.LimitReader(resp.Body, 64<<20))
	if err != nil {
		return nil, fmt.Errorf("ollama read response: %w", err)
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("ollama %d: %s", resp.StatusCode,
			string(payload[:min(300, len(payload))]))
	}

	var result struct {
		Embeddings [][]float64 `json:"embeddings"`
	}
	if err := json.Unmarshal(payload, &result); err != nil {
		return nil, fmt.Errorf("ollama decode: %w", err)
	}
	if len(result.Embeddings) != len(texts) {
		return nil, fmt.Errorf("ollama returned %d embeddings for %d inputs",
			len(result.Embeddings), len(texts))
	}
	return result.Embeddings, nil
}

func (e *EmbeddingClient) callTEI(ctx context.Context, texts []string) ([][]float64, error) {
	body := map[string]any{"inputs": texts, "truncate": true}
	data, err := json.Marshal(body)
	if err != nil {
		return nil, fmt.Errorf("encode tei request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, e.url+"/embed", bytes.NewReader(data))
	if err != nil {
		return nil, fmt.Errorf("create tei request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := e.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("tei embed: %w", err)
	}
	defer resp.Body.Close()

	payload, err := io.ReadAll(io.LimitReader(resp.Body, 64<<20))
	if err != nil {
		return nil, fmt.Errorf("tei read response: %w", err)
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("tei %d: %s", resp.StatusCode, string(payload[:min(300, len(payload))]))
	}

	var embeddings [][]float64
	if err := json.Unmarshal(payload, &embeddings); err != nil {
		return nil, fmt.Errorf("tei decode: %w", err)
	}
	return embeddings, nil
}

// generateLocal produces deterministic pseudo-embeddings (dev/test only).
func (e *EmbeddingClient) generateLocal(texts []string) [][]float64 {
	embeddings := make([][]float64, len(texts))
	for i, text := range texts {
		emb := make([]float64, e.dimensions)
		for j, b := range []byte(text) {
			emb[j%e.dimensions] += float64(b) / 1000.0
		}
		// L2 normalize
		norm := 0.0
		for _, v := range emb {
			norm += v * v
		}
		norm = math.Sqrt(norm)
		if norm > 0 {
			for j := range emb {
				emb[j] /= norm
			}
		}
		embeddings[i] = emb
	}
	return embeddings
}

func envOr(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

func envIntOr(key string, fallback int) int {
	if v := os.Getenv(key); v != "" {
		var i int
		fmt.Sscanf(v, "%d", &i)
		if i > 0 {
			return i
		}
	}
	return fallback
}

// init logs the embedding provider at startup.
func init() {
	provider := envOr("EMBEDDING_PROVIDER", "local")
	slog.Debug("Embedding client initialized", "provider", provider)
}
